# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Emmamems18/flyrank-ml-internship-my-submission-assignment1/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

Lane: Striking Distance Keyword Optimization
I am selecting the Striking Distance lane to identify high-potential content currently stuck on Page 2 or 3 of search results. Moving a page from position 11 – 20 to position 1 – 5 yields a massive jump in organic clicks.
By analyzing avg_position, position_tier, search_volume, and ctr across these 30,000 rows, this project will help SEO managers target low-hanging fruit for immediate optimization.

In [12]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
elif os.path.basename(os.getcwd()) == "notebooks":
    os.chdir("..")

import pandas as pd, numpy as np
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

In [13]:
df.shape

(30000, 44)

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 44 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   content_id              30000 non-null  object 
 1   client_id               30000 non-null  object 
 2   search_volume           27532 non-null  float64
 3   competition             27532 non-null  float64
 4   competition_level       27390 non-null  object 
 5   cpc                     27532 non-null  float64
 6   content_type            30000 non-null  object 
 7   main_intent             27626 non-null  object 
 8   word_count              22301 non-null  float64
 9   char_count              22301 non-null  float64
 10  provider_used           8562 non-null   object 
 11  model_used              24267 non-null  object 
 12  impressions_90d         30000 non-null  int64  
 13  clicks_90d              30000 non-null  int64  
 14  pageviews_90d           30000 non-null

In [15]:
# This cell is for CODE (numbers, a query, a check).
df.head()
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Unit of Analysis: A single piece of content (content_id).

Decision Improved: Deciding which specific "striking distance" pages (currently ranking on Page 2 or 3) should be prioritized for immediate SEO optimization resources.

Who Acts & Action Taken: The SEO Specialist or Content Manager. They will take direct action by updating on-page keywords, refreshing content depth, and building internal links for the flagged pages.

Cost of a Wrong Call:
False Positive (Recommending a low-potential page): Wastes valuable editorial hours and marketing budget on articles that fail to move up to Page 1.

False Negative (Missing a high-potential page): Leaves easy, high-intent organic traffic on the table, allowing competitors to capture Page 1 market share instead.

In [16]:
# This cell is for CODE (numbers, a query, a check).
# Check total content pieces vs striking distance opportunities (Page 2 & 3 rankings)
total_pages = len(df)
striking_distance_pages = df[
    df["position_tier"].isin(["page_2_3", "page_3_5"])
    | ((df["avg_position"] >= 11) & (df["avg_position"] <= 30))
]

count_striking = len(striking_distance_pages)
pct_striking = (count_striking / total_pages) * 100
avg_search_vol = striking_distance_pages["search_volume"].mean()

print(f"Total Pages in Dataset: {total_pages:,}")
print(
    f"Striking Distance Pages (Positions 11-30): {count_striking:,}"
    f" ({pct_striking:.1f}%)"
)
print(
    f"Average Monthly Search Volume for Striking Pages:"
    f" {avg_search_vol:.0f} searches"
)
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Total Pages in Dataset: 30,000
Striking Distance Pages (Positions 11-30): 13,488 (45.0%)
Average Monthly Search Volume for Striking Pages: 203 searches


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

Starter Dataset Key Metrics:

Total Content Assets: 30,000 pages.

Striking Distance Opportunity: 13,488 pages (45.0% of the catalog) rank in positions 11 – 30 (Pages 2 and 3).

Average Search Demand: Striking distance pages average 203 monthly searches each.

Why this lane is worth 7 weeks: Almost half the dataset represents low-hanging fruit sitting just off Page 1. An editorial team cannot manually review 13,488 candidates; machine learning is needed to systematically rank these pages so resources are spent only on high-yield targets.

In [17]:
# This cell is for CODE (numbers, a query, a check).
# Calculate key numbers for the Striking Distance lane
total_pages = len(df)
striking_pages = df[(df["avg_position"] >= 11) & (df["avg_position"] <= 30)]

count_striking = len(striking_pages)
pct_striking = (count_striking / total_pages) * 100
avg_search_vol = striking_pages["search_volume"].mean()

# Display summary
print(f"Total Pages in Dataset: {total_pages:,}")
print(
    f"Striking Distance Pages (Positions 11-30): {count_striking:,}"
    f" ({pct_striking:.1f}%)"
)
print(
    f"Average Monthly Search Volume for Striking Pages:"
    f" {avg_search_vol:.0f} searches"
)
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Total Pages in Dataset: 30,000
Striking Distance Pages (Positions 11-30): 10,201 (34.0%)
Average Monthly Search Volume for Striking Pages: 161 searches


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

What this work CAN claim:

Observed patterns: Identifies empirical relationships in historical data showing which page attributes co-occur with better ranking performance.

Directional guidance: Provides relative prioritization scores to rank the 10,201 striking-distance candidates from highest to lowest expected return on effort.

Decision-support: Reduces human effort by narrowing down thousands of unorganized pages into an actionable shortlist for editorial teams.

What this work CANNOT claim:

Causal proof: Observed correlations do not prove causality. Increasing a page's word count or updating its title does not guarantee Google will move it to Page 1.

Predicting Google's algorithm: Google uses hundreds of proprietary, real-time, and external signals (e.g., backlinks, user search history, domain authority) not present in this dataset.

Guaranteed outcomes: Models support decision-making but cannot guarantee specific position jumps or exact traffic gains.

In [18]:
# This cell is for CODE (numbers, a query, a check).
# Check correlations with average position to illustrate directional trends vs. noise
correlation_cols = [
    "avg_position",
    "word_count",
    "ctr",
    "days_since_last_update",
    "engagement_rate",
]
correlations = df[correlation_cols].corr()["avg_position"].drop("avg_position")

print("Correlation with Average Position:")
print(correlations.round(3))
print(
    "\nNote: Low to moderate correlation values show directional signals,"
    " not deterministic rules."
)
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Correlation with Average Position:
word_count                0.124
ctr                      -0.073
days_since_last_update    0.070
engagement_rate          -0.018
Name: avg_position, dtype: float64

Note: Low to moderate correlation values show directional signals, not deterministic rules.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.